### PR TITLE
[CARBONDATA-951] create table ddl can specify a bitmap option and add a new Encoding: BITMAP

### PR DESCRIPTION
**Pull request by mayunSaicmotor**
create table ddl can specify a bitmap option
add a new Encoding: BITMAP

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1731/

**Comment by QiangCai** [comment]
Regarding BITMAP encoding:
1.  the column should use DICTIONARY encoding also.
2.  the column will be a dimension, maybe sorted column or unsorted column

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1762/

**Comment by mayunSaicmotor** [comment]
fixed for the below items
1.the column should use DICTIONARY encoding also.
2.the column will be a dimension, maybe sorted column or unsorted column

**Comment by QiangCai** [comment]
Please check whether BITMAP column  use dictionary encoding or not.
If the column don't user dictionary encoding, throw error and stop flow.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1777/

File: format/src/main/thrift/carbondata.thrift [hunk 1]
```diff
@@ -130,6 +130,7 @@ struct DataChunk2{
     9: optional list<binary> encoder_meta; // Extra information required by encoders
     10: optional BlockletMinMaxIndex min_max; 
     11: optional i32 numberOfRowsInpage;
+    12: optional list<i32> bitmap_encoded_dictionaries;
```
Comment (QiangCai) [inline format/src/main/thrift/carbondata.thrift]: What's the purpose?

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 2]
```diff
@@ -273,7 +273,7 @@ class AlterTableProcessor(
     columnSchema.setDataType(dataType)
     columnSchema.setColumnName(colName)
     if (alterTableModel.highCardinalityDims.contains(colName)) {
-      encoders.remove(encoders.remove(Encoding.DICTIONARY))
+      encoders.remove(Encoding.DICTIONARY)
```
Comment (QiangCai) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: please keep old code, avoid conflict when rebase master

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 3]
```diff
@@ -366,6 +366,12 @@ class TableNewProcessor(cm: TableModel) {
     val LOGGER = LogServiceFactory.getLogService(TableNewProcessor.getClass.getName)
     var allColumns = Seq[ColumnSchema]()
     var index = 0
+    var bitmapStr: String = cm.tableProperties
+    .get(CarbonCommonConstants.BITMAP_ENCODING).getOrElse(null)
```
Comment (QiangCai) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: please use Option directly.
change bitmapStr != null to bitmapOption.isDefine

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 4]
```diff
@@ -385,6 +391,13 @@ class TableNewProcessor(cm: TableModel) {
         field.scale,
         field.schemaOrdinal)
       columnSchema.setSortColumn(true)
+      if (bitmapCols.contains(field.column)) {
+        if (columnSchema.getEncodingList.contains(Encoding.DICTIONARY)) {
```
Comment (QiangCai) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: please use columnSchema.hasEncoding(Encoding encoding) method

Comment (QiangCai) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: why not add these code into getColumnSchema method?

File: examples/spark/src/main/scala/org/apache/carbondata/examples/CarbonBitMapEncodingExample.scala [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.examples
+
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.util.CarbonProperties
+import org.apache.carbondata.examples.util.ExampleUtils
+
+object CarbonBitMapEncodingExample {
```
Comment (QiangCai) [inline examples/spark/src/main/scala/org/apache/carbondata/examples/CarbonBitMapEncodingExample.scala]: please add test case to check function also.
Example is just a demo, will not be run during building.

File: format/src/main/thrift/carbondata.thrift (hunk 1 above)
Comment (mayunSaicmotor) [inline format/src/main/thrift/carbondata.thrift]: store each chunk' dictionaries, for example if in 2 chunks one, have dictionary 2 and 3, and the other have 3 and 4. There are different between the 2 chunks. So bitmap_encoded_dictionaries to store the dictionaries for each chunk.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1846/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1847/

File: examples/spark/src/main/scala/org/apache/carbondata/examples/CarbonBitMapEncodingExample.scala (hunk 5 above)
Comment (mayunSaicmotor) [inline examples/spark/src/main/scala/org/apache/carbondata/examples/CarbonBitMapEncodingExample.scala]: added just test case please review

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1848/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1873/

**Pull request closed without merging**